middle_end/flambda2/types/grammar/type_grammar.ml

```diff
@@ -1451,6 +1498,367 @@ and remove_unused_closure_vars_env_extension ({ equations } as env_extension)
+    let ty' = project_variables_out ~to_project ~expand ty in
+    if ty == ty' then head else Boxed_int32 ty'
+  | Boxed_int64 ty ->
+    let ty' = project_variables_out ~to_project ~expand ty in
+    if ty == ty' then head else Boxed_int64 ty'
+  | Boxed_nativeint ty ->
+    let ty' = project_variables_out ~to_project ~expand ty in
+    if ty == ty' then head else Boxed_nativeint ty'
+  | Closures { by_closure_id } ->
+    let by_closure_id' =
+      project_row_like_for_closures ~to_project ~expand by_closure_id
+    in
+    if by_closure_id == by_closure_id'
+    then head
+    else Closures { by_closure_id = by_closure_id' }
+  | String _ -> head
+  | Array { element_kind; length } ->
+    let length' = project_variables_out ~to_project ~expand length in
+    if length == length' then head else Array { element_kind; length = length' }
+
+and project_head_of_kind_naked_immediate ~to_project ~expand head =
+  match head with
+  | Naked_immediates _ -> head
+  | Is_int ty ->
+    let ty' = project_variables_out ~to_project ~expand ty in
+    if ty == ty' then head else Is_int ty'
+  | Get_tag ty ->
+    let ty' = project_variables_out ~to_project ~expand ty in
+    if ty == ty' then head else Get_tag ty'
+
+and project_head_of_kind_naked_float ~to_project:_ ~expand:_ head = head
+
+and project_head_of_kind_naked_int32 ~to_project:_ ~expand:_ head = head
+
+and project_head_of_kind_naked_int64 ~to_project:_ ~expand:_ head = head
+
+and project_head_of_kind_naked_nativeint ~to_project:_ ~expand:_ head = head
+
+and project_head_of_kind_rec_info ~to_project ~expand:_ head =
+  match (head : head_of_kind_rec_info) with
+  | Const _ | Succ _ | Unroll_to _ -> head
+  | Var var ->
+    if not (Variable.Set.mem var to_project)
+    then head
+    else Misc.fatal_error "Project of depth variables is not implemented"
+
+and project_row_like_for_blocks ~to_project ~expand
+    ({ known_tags; other_tags } as blocks) =
+  let known_tags' =
+    Tag.Map.map_sharing
+      (fun ({ index; maps_to; env_extension } as case) ->
+        let env_extension' =
+          project_env_extension ~to_project ~expand env_extension
+        in
+        let maps_to' =
+          project_int_indexed_product ~to_project ~expand maps_to
+        in
+        if env_extension == env_extension' && maps_to == maps_to'
+        then case
+        else { index; env_extension = env_extension'; maps_to = maps_to' })
+      known_tags
+  in
+  let other_tags' : _ Or_bottom.t =
+    match other_tags with
+    | Bottom -> Bottom
+    | Ok { index; maps_to; env_extension } ->
+      let env_extension' =
+        project_env_extension ~to_project ~expand env_extension
+      in
+      let maps_to' = project_int_indexed_product ~to_project ~expand maps_to in
+      if env_extension == env_extension' && maps_to == maps_to'
+      then other_tags
+      else Ok { index; env_extension = env_extension'; maps_to = maps_to' }
+  in
+  if known_tags == known_tags' && other_tags == other_tags'
```

Comment:
It could look like this maybe

```OCaml
module Share : sig

  type 'a map = ('a -> 'a) * 'a
  type 'a ret

  val bind : 'a map -> ('a -> 'b ret) -> 'b ret
  val (let*) : 'a map -> ('a -> 'b ret) -> 'b ret
  val return : 'a -> 'a ret
  val select : 'a -> 'a ret -> 'a

end = struct

  type 'a map = ('a -> 'a) * 'a
  type 'a ret = 'a * bool

  let bind (type t) ((f : t -> t), (v : t)) r =
    let v' = f v in
    let b, eq = r v' in
    b, eq && v' == v

  let (let*) = bind
  let return v = v, true
  let select a (a', c) =
    if c then a else a'

end

let fa x = x+1
let fb x = not x

let f ((a, b, c) as orig) =
  let open Share in
  select orig @@
  let* a = fa, a in
  let* b = fb, b in
  return (a, b, c)
```